src/controller/index.ts

```diff
@@ -404,6 +408,67 @@ export default class SimulariumController {
         return this.playBackFile;
     }
 
+    public async getMetrics(config: NetConnectionParams): Promise<void> {
+        if (
+            !this.metricsCalculator ||
+            !this.metricsCalculator.socketIsValid()
+        ) {
+            const webSocketClient = new WebsocketClient(config, this.onError);
+            this.metricsCalculator = new RemoteMetricsCalculator(
+                webSocketClient,
+                this.onError
+            );
+            await this.metricsCalculator.connectToRemoteServer();
+        }
+        this.metricsCalculator.getAvailableMetrics();
+    }
+
+    public async getPlotData(config: NetConnectionParams): Promise<void> {
+        if (!this.simulator || this.simulator instanceof ClientSimulator) {
+            // getting plot data requires RemoteSimulator or LocalFileSimulator
+            return;
+        }
+
+        if (
+            !this.metricsCalculator ||
+            !this.metricsCalculator.socketIsValid()
+        ) {
+            const webSocketClient =
+                this.remoteWebsocketClient &&
+                this.remoteWebsocketClient.socketIsValid()
+                    ? this.remoteWebsocketClient
+                    : new WebsocketClient(config, this.onError);
+            this.metricsCalculator = new RemoteMetricsCalculator(
+                webSocketClient,
+                this.onError
+            );
+        }
+
+        // TODO: add form for users to select the metrics of hardcoding them
+        const plot1 = {
+            plotType: "scatter",
+            metricsIdx: 0,
+            metricsIdy: 2,
+            scatterPlotMode: "lines",
+        };
+        const plot2 = {
+            plotType: "histogram",
+            metricsIdx: 3,
+        };
+
+        if (this.simulator instanceof LocalFileSimulator) {
```

Comment:
On second thought, there is probably some way to set up the simulariumFile in the metrics calculator earlier before this call happens?

Reply:
like preemptively send the simulariumFile to the metrics calculator before the user requests any plot data?

it is certainly really expensive to send the whole simulariumFile to the metricsCalculator, which will be living on an EC2 instance with Octopus and Autoconversion, but my understanding is that it's needed in order to do the calculations.

It definitely feels like there's a smarter way to do this, at least in some cases. Like for the case of people adding plots to files that they had just converted via autoconversion, maybe we could cache them on the server somewhere so that we wouldn't have to send the file back and forth for no reason again? But also, we probably don't want other people's large autoconverted files taking up a bunch of space on the server...

I'm spinning in circles. I'll look at this again in the morning, perhaps I'll dream up a solution

Reply:
The whole idea of sending all the trajectory data up to a remote server just makes me think we need to check to see if the data is already local on the server, or else tell users that it might take a while.  How long does it take in practice with the file sizes you have?

Reply:
Yeah for a 90 MB file, it takes about 30 seconds to do the actual metrics calculation, so for a local file it takes more like a few minutes, since we have to send the whole file back and forth. But it would probably be a good idea to cache the file on the server when someone sends a request for a plot that includes the whole trajectory, since there's probably a good chance that they'll request another plot and it'd be silly to go through the whole file exchange again